### PR TITLE
nixos/localsend: init

### PR DESCRIPTION
**Pull request by Luis-Licea**
## Description of changes

Create a module for localsend providing the option to open the TCP firewall 53317.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/programs/localsend.nix
Comment (Aleksanaa) [inline nixos/modules/programs/localsend.nix:12]:
```suggestion
    enable = mkEnableOption "localsend, an open source cross-platform alternative to AirDrop";
```

note that `mdDoc` is removed. You should remove it altogether (description is mdDoc by default now)

File: nixos/modules/programs/localsend.nix
Comment (Aleksanaa) [inline nixos/modules/programs/localsend.nix:24]:
```suggestion
    environment.systemPackages = with pkgs; [ localsend ];
```

Looks like you are using alejandra? If you want to format, format with `nixfmt-rfc-style`.

**Review by Aleksanaa — COMMENTED** [review]
``

**Comment by Aleksanaa** [comment]
Also please reword to `nixos/localsend: init`

File: nixos/modules/programs/localsend.nix
Comment (Luis-Licea) [reply nixos/modules/programs/localsend.nix]: Good point. Changed to `true`.

File: nixos/modules/programs/localsend.nix
Comment (Luis-Licea) [reply nixos/modules/programs/localsend.nix]: Removed.

File: nixos/modules/programs/localsend.nix
Comment (Luis-Licea) [reply nixos/modules/programs/localsend.nix]: My system is not state version `24.05` yet so I cannot install it. Is using `nixfmt` acceptable?

**Comment by Luis-Licea** [comment]
I have a question. I carefully wrote the test `nixos/tests/localsend.nix` but I have no idea how to execute it or if it works. Will CI take care of executing it for me?

**Comment by Luis-Licea** [comment]
I got the following error after I ran `nix build .#nixosTests.localsend`:
```
error: The option `nodes.machine.programs.localsend' does not exist. Definition values:
       - In `makeTest parameters':
           {
             enable = true;
             openFirewall = true;
           }
```
I am not sure why, but `nodes.machine.programs` does not have access to the `localsend` options I defined.

**Comment by h7x4** [comment]
You'll need to add your module to `nixos/modules/module-list`

**Comment by Luis-Licea** [comment]
Adding `localsend` to `nixos/modules/module-list.nix` solved that specific program. Now the problem is that `localsend` does not get installed into the `nodes.machine`.

The line `machine.execute("localsend")` in the test script `tests/localsend.nix` returns `machine # bash: line 1: localsend: command not found`.  I also tried replacing the line with `machine.execute("${pkgs.localsend}/bin/localsend")`, but this did not work either.

**Comment by Luis-Licea** [comment]
The test is fixed. Running `nix build .#nixosTests.localsend` succeeds.

File: nixos/modules/programs/localsend.nix
Comment (h7x4) [inline nixos/modules/programs/localsend.nix:16]:
```suggestion
        Whether to open the firewall port ${toString firewallPort} for receiving files.
```

File: nixos/modules/programs/localsend.nix
Comment (h7x4) [inline nixos/modules/programs/localsend.nix:7]: Maybe you could try to avoid file-wide `with lib` statements? Either inheriting the functions you are using, or accessing them directly using `lib.<fun>` would be better.

File: nixos/modules/module-list.nix
Comment (h7x4) [inline nixos/modules/module-list.nix:225]: Could you leave this sort in another commit, or for another PR? It's not really relevant for localsend.

Comment (Luis-Licea) [reply nixos/modules/module-list.nix]: The sort has been reverted.

File: nixos/modules/programs/localsend.nix
Comment (Luis-Licea) [reply nixos/modules/programs/localsend.nix]: Changed `with lib` to `lib.<fun>`.

**Comment by Luis-Licea** [comment]
This pull request is ready to merge.

**Comment by Aleksanaa** [comment]
@ofborg test nixosTests.localsend

**Comment by JohnRTitor** [comment]
Hi there @Pandapip1, I have cherry-picked your changes here. Please help review this!

**Comment by JohnRTitor** [comment]
I have reworded the commits and applied `nixfmt-rfc-style` as appropriate.

File: nixos/modules/programs/localsend.nix
Comment (Pandapip1) [inline nixos/modules/programs/localsend.nix:28]:
```suggestion

  meta.maintainers = pkgs.localsend.meta.maintainers;
}
```

File: pkgs/applications/networking/localsend/default.nix
Comment (Pandapip1) [inline pkgs/applications/networking/localsend/default.nix:101]: Probably should add @Luis-Licea here too

File: nixos/tests/localsend.nix
Comment (Pandapip1) [inline nixos/tests/localsend.nix:9]: Consider using Wayland instead?

Comment (Luis-Licea) [reply nixos/tests/localsend.nix:9]: Does the Wayland module provide a graphical interface that we can wait for? The test relies on the command `machine.wait_for_window("LocalSend", 10)`. If the same functionality can be tested using Wayland, then I would also prefer to use Wayland instead of X11 for the test, but if I remember correctly, the Wayland module sets up a headless VM.

I do not have electicity at home at the moment so I will not be able to test using Wayland instead of X11 for at least a few days.

File: nixos/modules/programs/localsend.nix
Comment (JohnRTitor) [reply nixos/modules/programs/localsend.nix:28]: If they are willing to maintain it of course, else @Luis-Licea or @Pandapip1 should be added.

File: nixos/tests/localsend.nix
Comment (JohnRTitor) [reply nixos/tests/localsend.nix:9]: > Does the Wayland module provide a graphical interface that we can wait for? The test relies on the command `machine.wait_for_window("LocalSend", 10)`. If the same functionality can be tested using Wayland, then I would also prefer to use Wayland instead of X11 for the test, but if I remember correctly, the Wayland module sets up a headless VM.

Exactly my point. I do not see it as a requirement to merge, but if @Pandapip1 would like to send a patch I am happy to commit it for you.

Comment (Pandapip1) [reply nixos/tests/localsend.nix:9]: Oh, I definitely don't see it as a requirement either. Certainly, it's not blocking.

File: nixos/modules/programs/localsend.nix
Comment (JohnRTitor) [reply nixos/modules/programs/localsend.nix:26]: Added you as the maintainer, as Luis is not present in the maintainers-list.nix.

**Review by Pandapip1 — APPROVED** [review]
LGTM

**Review by JohnRTitor — APPROVED** [review]
Self assigned so I do not lose track of this. Will merge once eval is completed.

**Comment by Pandapip1** [comment]
@JohnRTitor all checks have passed

**Pull request merged**
